nep-sales.mediawiki

```diff
@@ -0,0 +1,123 @@
+
+==Specification==
+
+We propose the following methods to be implemented in any crowdsale contract that requires investor verification as well as sales restricted to a timeframe.
+
+The methods starting with "kyc" are methods for individuals, indicating their status in the current sale.
+
+The methods starting with "sale" are methods for the entire sale, indicating the status of the current sale.
+
+This standard defines two method types:
+
+* '''(Required)''' : methods that are compulsory in order to be NEP-compliant. Methods are required unless marked otherwise. These methods are mostly user-facing and thus must be standardised for a smooth experience.
+
+* '''(Optional)''' : methods that are optionally implemented. These methods are not all user-facing and contains certains parameters not present in all sales. However, developer tools are encouraged to follow this specification for the best experience.
+
+===Methods===
+
+====mintTokens====
+
+* Syntax: <code>public static string mintTokens(byte[] address)</code>
+
+* Remarks: ''mintToken'' attempts to create tokens for the specified address. If successful, this action should increase the totalSupply of the tokens. This method must be implemented together with the ''token_mint'' event.
+
+====kycStatus====
+
+* Syntax: <code>public static bool kycStatus(byte[] address)</code>
+
+* Remarks: ''kycStatus'' Checks the whitelist status for the provided address. Returns True if registered.
+
+====kycLimit====
+
+* Syntax: <code>public static int kycLimit(byte[] address)</code>
+
+* Remarks: ''kycLimit'' returns the maximum amount of tokens that the provided address can exchange for. This method should ignore the global remaining tokens and focus on the personal limit. Sales without a personal limit should return the largest possible integer.
+
+* Example: For an address registered to exchange up to 1000 tokens, in a sale where 1 NEO is worth 100 tokens, 1 GAS is worth 10 tokens, this method should return 1000 at the beginning. After contributing 1 NEO and 10 GAS, this method should return 800. 
+
+====saleStatus====
+
+* Syntax: <code>public static bool saleStatus()</code>
+
+* Remarks: ''saleStatus'' returns the current state of the sale. Returns true if sale is active.
+
+====saleRemaining====
+
+* Syntax: <code>public static int saleRemaining()</code>
+
+* Remarks: "saleRemaining" returns the amount of tokens left for sale. This is the remaining amount of token left available for exchange. Sales without a limit should return the largest possible integer.
+
+====saleDetails ''(optional)''====
+
+* Syntax: <code>public static string saleDetails()</code>

```

Comment:
It might be cool to add an optional `byte[] address` argument that could be used to generate a personalized status message for a given address. This message could indicate any per-address sale details (e.g. if different users get different token ratios or have different limits). This would give ICO implementors a bit more flexibility and power in delivering a more useful and accurate message in the wallet interface. Worth consideration, anyway.

Reply:
Ok, by doing this, we can remove the limit / sale remaining by letting the contract owner compile their own custom string that may target the specific address if needed.